sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusListener.cs

```diff
@@ -550,6 +591,262 @@ private async Task RunBatchReceiveLoopAsync(CancellationToken cancellationToken)
+                            _isSessionsEnabled ? Constants.ProcessSessionMessagesActivityName : Constants.ProcessMessagesActivityName,
+                            DiagnosticScope.ActivityKind.Consumer,
+                            MessagingDiagnosticOperation.Process);
+
+            scope.SetMessageData(messagesArray);
+
+            scope.Start();
+            FunctionResult result = await _triggerExecutor.TryExecuteAsync(input.GetTriggerFunctionData(), cancellationToken).ConfigureAwait(false);
+            if (result.Exception != null)
+            {
+                scope.Failed(result.Exception);
+            }
+            receiveActions.EndExecutionScope();
+
+            var processedMessages = messagesArray.Concat(receiveActions.Messages.Keys);
+            // Complete batch of messages only if the execution was successful
+            if (_autoCompleteMessages && result.Succeeded)
+            {
+                List<Task> completeTasks = new List<Task>();
+                foreach (ServiceBusReceivedMessage message in processedMessages)
+                {
+                    // Skip messages that were settled in the user's function
+                    if (input.MessageActions.SettledMessages.ContainsKey(message))
+                    {
+                        continue;
+                    }
+
+                    // Pass CancellationToken.None to allow autocompletion to finish even when shutting down
+                    completeTasks.Add(receiver.CompleteMessageAsync(message, CancellationToken.None));
+                }
+
+                await Task.WhenAll(completeTasks).ConfigureAwait(false);
+            }
+            else if (!result.Succeeded)
+            {
+                // For failed executions, we abandon the messages regardless of the autoCompleteMessages configuration.
+                // This matches the behavior that happens for single dispatch functions as the processor does the same thing
+                // in the Service Bus SDK.
+
+                List<Task> abandonTasks = new();
+                foreach (ServiceBusReceivedMessage message in processedMessages)
+                {
+                    // skip messages that were settled in the user's function
+                    if (input.MessageActions.SettledMessages.ContainsKey(message))
+                    {
+                        continue;
+                    }
+
+                    // Pass CancellationToken.None to allow abandon to finish even when shutting down
+                    abandonTasks.Add(receiver.AbandonMessageAsync(message, cancellationToken: CancellationToken.None));
+                }
+
+                await Task.WhenAll(abandonTasks).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<IReadOnlyList<ServiceBusReceivedMessage>> ReceiveMessages(ServiceBusReceiver receiver, TimeSpan? maxWaitTime, CancellationTokenSource cancellationTokenSource)
```

Comment:
Why does this need to be a separate method?

Reply:
I was calling it multiple times at one point. Fixed.